vyper/old_codegen/parser.py

```diff
@@ -177,8 +178,46 @@ def parse_regular_functions(
     ]
     runtime.extend(internal_funcs)
 
-    # TODO CMC 20210911 why does the lll have a trailing 0
-    o.append(["return", 0, ["lll", runtime, 0]])
+    immutables = [_global for _global in global_ctx._globals.values() if _global.is_immutable]
+
+    # TODO: enable usage of the data section beyond just user defined immutables
+    # https://github.com/vyperlang/vyper/pull/2466#discussion_r722816358
+    if len(immutables) > 0:
+        # find position of the last immutable so we do not overwrite it in memory
+        # when we codecopy the runtime code to memory
+        immutables = sorted(immutables, key=lambda imm: imm.pos)
+        start_pos = immutables[-1].pos + immutables[-1].size * 32
+        # create sequence of actions to copy immutables to the end of the runtime code in memory
+        data_section = []
+        for immutable in immutables:
+            # store each immutable at the end of the runtime code
+            memory_loc, offset = (
+                immutable.pos,
+                immutable.data_offset,
+            )
+            lhs = LLLnode.from_list(
+                ["add", start_pos + offset, "_lllsz"], typ=immutable.typ, location="memory"
+            )
+            rhs = LLLnode.from_list(memory_loc, typ=immutable.typ, location="memory")
+            data_section.append(make_setter(lhs, rhs, None))
+
+        data_section_size = sum([immutable.size * 32 for immutable in immutables])
+        o.append(
+            [
+                "with",
+                "_lllsz",  # keep size of runtime bytecode in sz var
+                ["lll", start_pos, runtime],  # store runtime code at `start_pos`
+                # sequence of copying immutables, with final action of returning the runtime code
+                ["seq", *data_section, ["return", start_pos, ["add", data_section_size, "_lllsz"]]],
+            ]
+        )
+
+    else:
+        # NOTE: lll macro trailing 0 is the location in memory to store
```

Comment:
just noticed this comment is outdated - should be `lll macro first argument is the location in memory to store`

Reply:
```suggestion
        # NOTE: lll macro first argument is the location in memory to store
```